### PR TITLE
Optimize the logic of finding cluster service/pod ip range

### PR DESCRIPTION
**Pull request by unilinu**
#### What type of PR is this?

Fix or Optimize

#### What this PR does / why we need it:

According to the original logic, do some optimizations:

1. Correct the label selector of kube-proxy from `component=kube-proxy` to `k8s-app=kube-proxy`

2. Similar to `findPodIPRange` function, add the redundant finding from kube-controller-manager.

### CODE REVIEW

File: pkg/discovery/network/generic.go [hunk 1]
```diff
@@ -78,22 +77,31 @@ func discoverNetwork(ctx context.Context, client controllerClient.Client) (*Clus
 
 func findClusterIPRange(ctx context.Context, client controllerClient.Client) (string, error) {
 	clusterIPRange, err := findClusterIPRangeFromApiserver(ctx, client)
-	if err != nil || clusterIPRange != "" {
-		return clusterIPRange, err
+	if clusterIPRange != "" {
```
Comment (tpantelis) [inline pkg/discovery/network/generic.go]: Why did you remove the error checking here and other places below? This hides the error which isn't best practice. An error here means it could not determine if the IP range could be obtained, most likely because the API server wasn't available, in which case we want to abort and report the error.

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/discovery/network/generic.go (hunk 1 above)
Comment (unilinu) [reply pkg/discovery/network/generic.go]: Well, It seems I understand wrong the purpose of these error checking. I thought this function is to detect the available ClusterIP range, but it seems to detect the availability of component/service related to ClusterIP. I will cancel the related modifications later. 

In addition, I think the point 1&2 should be retained, but for point 3, I wonder if add the redundant finding from kube-controller-manager.

Comment (unilinu) [reply pkg/discovery/network/generic.go]: @tpantelis Another question, the original if conditions `err != nil || clusterIPRange != ""` is almost true, which will cover/skip the following judging!!
In my opinion, this function is used to find the service cluster ip range. Even if occurs error, which should be detected and handled somewhere, adding a waring/error log but continuing finding is a good idea here.

If you agree, I will modify it as above. If not, looking forward to your reply.

Comment (tpantelis) [reply pkg/discovery/network/generic.go]: > In my opinion, this function is used to find the service cluster ip range. Even if occurs error, which should be detected and handled somewhere, adding a waring/error log but continuing finding is a good idea here.
>
> If you agree, I will modify it as above. If not, looking forward to your reply.

I don't agree. Consider this scenario:

Let's say in a particular environment the first call, `findClusterIPRangeFromApiserver`, under  normal circumstances with no error finds the cluster IP. But on a particular run a transient error occurs, eg the API server is busy and times out, in which case we would just log a warning and continue. For the rest of the code below, the API server responds and no error occurs but the cluster IP isn't found by any of the following methods. This function would return and report to the caller that no cluster IP was found and no error occurred. But this is incorrect as an error did occur but it was essentially ignored by just logging it. We must propagate errors to the caller to indicate that the result was indeterminate and to let the caller handle the error as it sees fit.

Comment (unilinu) [reply pkg/discovery/network/generic.go]: I can understand it. So, each error should be checked and not be covered by assignments. However, there is no doubt that `err != nil || clusterIPRange != ""` condition will cover/skip the following statements. This point must be corrected. I'll modify the PR correspondingly.

Comment (tpantelis) [reply pkg/discovery/network/generic.go]: No - the error checking was fine as is. Given your recent changes and the scenario I outlined above, it would return a valid cluster IP **and** an error which is non-sensical and non-idiomatic.

Please just focus on the purpose of this PR as outlined in your summary and not modify the existing error handling.

**Comment by unilinu** [comment]
Thanks for your patience firstly. In fact, I wonder why the error checking was fine. Is an error or a valid cluster ip always return true? Thus, what effect the following methods make?在 2024年12月3日，20:48，Thomas Pantelis ***@***.***> 写道：﻿
@tpantelis commented on this pull request.



In pkg/discovery/network/generic.go:
> @@ -78,22 +77,31 @@ func discoverNetwork(ctx context.Context, client controllerClient.Client) (*Clus
 
 func findClusterIPRange(ctx context.Context, client controllerClient.Client) (string, error) {
 	clusterIPRange, err := findClusterIPRangeFromApiserver(ctx, client)
-	if err != nil || clusterIPRange != "" {
-		return clusterIPRange, err
+	if clusterIPRange != "" {

No - the error checking was fine as is. Given your recent changes and the scenario I outlined above, it would return a valid cluster IP and an error which is non-sensical and non-idiomatic.

File: pkg/discovery/network/generic.go (hunk 1 above)
Comment (unilinu) [reply pkg/discovery/network/generic.go]: Thanks for your patience. I know what I misunderstood. `err != nil || clusterIPRange != ""` is not always true. There is a situation not finding cluster ip range from the args of apiserver/controller-manager. 

I have modified my patch, removing the changes of error checking.

File: pkg/discovery/network/generic.go [hunk 2]
```diff
@@ -153,50 +162,23 @@ func parseServiceCIDRFrom(msg string) (string, error) {
...
-	return parseToPodCidr(nodes.Items)
-}
-
-func parseToPodCidr(nodes []corev1.Node) (string, error) {
-	// In K8s, each node is typically assigned a unique PodCIDR range for the pods that run on that node.
-	// Each node's PodCIDR is used to allocate IP addresses to the pods scheduled on that node. Only if
-	// the cluster is a single node deployment, we should rely on the node.Spec.PodCIDR as podCIDR of the cluster.
-	if len(nodes) == 1 {
-		return nodes[0].Spec.PodCIDR, nil
-	}
```
Comment (tpantelis) [inline pkg/discovery/network/generic.go]: From the PR summary:

> The node.spec.podCIDR field is not the cluster pod ip range, but the node-assigned.

But this is only considered if there's only one node as explained in the comments above and if the pod CIDR wasn't found from any prior methods. It seems valid to keep this.

**Review by tpantelis — CHANGES_REQUESTED** [review]

**Comment by tpantelis** [comment]
@unilinu The DCO job failed - you need to add Signed-off-by your commit.

File: pkg/discovery/network/generic.go (hunk 2 above)
Comment (unilinu) [reply pkg/discovery/network/generic.go]: You are right. I have fixed it.

**Review by tpantelis — APPROVED** [review]

**Review by tpantelis — CHANGES_REQUESTED** [review]
Please add unit tests to cover the new code you added.

**Comment by tpantelis** [comment]
[Linting/Commit Message](https://github.com/submariner-io/submariner-operator/actions/runs/12154031393/job/33911594332?pr=3285) job failed.

Also unit tests failed.

**Comment by unilinu** [comment]
> [Linting/Commit Message](https://github.com/submariner-io/submariner-operator/actions/runs/12154031393/job/33911594332?pr=3285) job failed.
> 
> Also unit tests failed.

Added and fixed.

File: pkg/discovery/network/generic_test.go [hunk 3]
```diff
@@ -120,30 +120,33 @@ var _ = Describe("Generic Network", func() {
...
-		})
 	})
 
 	When("There is a kube-controller pod with the right parameter passed as an Arg", func() {
 		BeforeEach(func(ctx SpecContext) {
 			clusterNet = testDiscoverGenericWith(
 				ctx,
-				fakePodWithArg("kube-controller-manager", []string{"kube-controller-manager"}, "--cluster-cidr="+testPodCIDR),
+				fakePodWithArg("kube-controller-manager", []string{"kube-controller-manager"},
+					[]string{"--cluster-cidr=" + testPodCIDR, "--service-cluster-ip-range=" + testServiceCIDR}),
```
Comment (tpantelis) [inline pkg/discovery/network/generic_test.go:138]: Should be able to use:

```suggestion
				fakeKubeControllerManagerPod(),
```

File: pkg/discovery/network/generic_test.go [hunk 4]
```diff
@@ -120,30 +120,33 @@ var _ = Describe("Generic Network", func() {
 			Expect(clusterNet).NotTo(BeNil())
 		})
 
-		It("Should return the ClusterNetwork structure with pod CIDR", func() {
+		It("Should return the ClusterNetwork structure with all CIDR", func() {
```
Comment (tpantelis) [inline pkg/discovery/network/generic_test.go]:
```suggestion
		It("Should return the ClusterNetwork structure with both CIDRs", func() {
```

File: pkg/discovery/network/generic_test.go [hunk 5]
```diff
@@ -120,30 +120,33 @@ var _ = Describe("Generic Network", func() {
...
 				ctx,
-				fakePodWithArg("kube-controller-manager", []string{"kube-controller-manager"}, "--cluster-cidr="+testPodCIDR),
+				fakePodWithArg("kube-controller-manager", []string{"kube-controller-manager"},
+					[]string{"--cluster-cidr=" + testPodCIDR, "--service-cluster-ip-range=" + testServiceCIDR}),
 			)
 			Expect(clusterNet).NotTo(BeNil())
 		})
 
-		It("Should return the ClusterNetwork structure with pod CIDR", func() {
+		It("Should return the ClusterNetwork structure with all CIDR", func() {
```
Comment (tpantelis) [inline pkg/discovery/network/generic_test.go]:
```suggestion
		It("Should return the ClusterNetwork structure with both CIDRs", func() {
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

**Comment by tpantelis** [comment]
[Linting/Commit Message](https://github.com/submariner-io/submariner-operator/actions/runs/12174365280/job/33973867105?pr=3285) job is still failing. The commit title is too long.

There's also [Go linting errors](https://github.com/submariner-io/submariner-operator/actions/runs/12174365280/job/33973867474?pr=3285).

**Comment by dfarrell07** [comment]
Thanks for the PR! Please check the list of CI jobs and see the details of the ones that failed.

(posted at the same time as Tom above, sorry for the repeat)

File: pkg/discovery/network/generic_test.go (hunk 3 above)
Comment (unilinu) [reply pkg/discovery/network/generic_test.go:138]: The test case is "There is a kube-controller pod with the right parameter **passed as an Arg**". The `fakeKubeControllerManagerPod` function should be and have been used in the last one naming "There is a kube-controller pod with the right parameter".

**Comment by unilinu** [comment]
> [Linting/Commit Message](https://github.com/submariner-io/submariner-operator/actions/runs/12174365280/job/33973867105?pr=3285) job is still failing. The commit title is too long.
> 
> There's also [Go linting errors](https://github.com/submariner-io/submariner-operator/actions/runs/12174365280/job/33973867474?pr=3285).

Thans for your patience. The errors you mentioned have been fixed.

**Review by tpantelis — APPROVED** [review]

**Review by skitt — APPROVED** [review]
